src/server/game/Handlers/MiscHandler.cpp

```diff
@@ -1117,36 +1117,56 @@ void WorldSession::HandleInspectHonorStatsOpcode(WorldPacket& recv_data)
     SendPacket(&data);
 }
 
-void WorldSession::HandleWorldTeleportOpcode(WorldPacket& recv_data)
+void WorldSession::HandleWorldTeleport(WorldPacket& message)
 {
-    uint32 time;
-    uint32 mapid;
-    float PositionX;
-    float PositionY;
-    float PositionZ;
-    float Orientation;
+    uint32_t    continentID;
+    uint32_t    requestTimeMs;
+    float       facing;
+    uint64_t    player;
+    Position    position;
+    Player*     pPlayer;
 
-    recv_data >> time;                                      // time in m.sec.
-    recv_data >> mapid;
-    recv_data >> PositionX;
-    recv_data >> PositionY;
-    recv_data >> PositionZ;
-    recv_data >> Orientation;                               // o (3.141593 = 180 degrees)
-
-    LOG_DEBUG("network", "WORLD: Received CMSG_WORLD_TELEPORT");
-
-    if (GetPlayer()->IsInFlight())
+    if (!AccountMgr::IsGMAccount(GetSecurity()))
     {
-        LOG_DEBUG("network", "Player '{}' ({}) in flight, ignore worldport command.", GetPlayer()->GetName(), GetPlayer()->GetGUID().ToString());
+        SendNotification(LANG_YOU_NOT_HAVE_PERMISSION);
         return;
     }
 
-    LOG_DEBUG("network", "CMSG_WORLD_TELEPORT: Player = {}, Time = {}, map = {}, x = {}, y = {}, z = {}, o = {}", GetPlayer()->GetName(), time, mapid, PositionX, PositionY, PositionZ, Orientation);
+    message >> requestTimeMs;
+    message >> continentID;
+    message >> player;
+    message >> position.m_positionX;
+    message >> position.m_positionY;
+    message >> position.m_positionZ;
+    message >> facing;
 
-    if (AccountMgr::IsAdminAccount(GetSecurity()))
-        GetPlayer()->TeleportTo(mapid, PositionX, PositionY, PositionZ, Orientation);
+    if (!player)
+    {
+        // CLIENT SENDS NULL GUID IF THE WORLDPORT COMMAND IS ISSUED FOR ITSELF
+        pPlayer = _player;
+    }
     else
-        SendNotification(LANG_YOU_NOT_HAVE_PERMISSION);
+    {
+        // ATTEMPT TO FIND THE PLAYER WITH THE PROVIDED GUID IN-GAME OR TELEPORT IT OFFLINE
+        pPlayer = ObjectAccessor::FindPlayer((ObjectGuid(player)));
+        if (!pPlayer)
+        {
+            QueryResult result = CharacterDatabase.Query("SELECT guid FROM characters WHERE guid={}", player);
+            if (result->GetRowCount())
+            {
+                uint32_t zoneID = MapMgr::instance()->GetZoneId(0, continentID, position.m_positionX, position.m_positionY, position.m_positionZ);
+
+                CharacterDatabase.Execute(
+                    "UPDATE characters SET map={}, position_x={}, position_y={}, position_z={}, orientation={}, zone={} WHERE guid={}",
+                    continentID, position.m_positionX, position.m_positionY, position.m_positionZ, facing, zoneID, player
```

Comment:
use TeleportTo() instead

Reply:
I'm explicitly porting the player offline. Want me to split the logic into a new function?